client/cli/src/runner.rs

```diff
@@ -259,24 +249,18 @@ impl<C: SubstrateCli> Runner<C> {
 		}
 	}
 
-	fn run_service_until_exit<T, F>(mut self, service_builder: F) -> Result<()>
-	where
-		F: FnOnce(Configuration) -> std::result::Result<T, sc_service::error::Error>,
-		T: AbstractService + Unpin,
-	{
-		let service = service_builder(self.config)?;
-
-		// we eagerly drop the service so that the internal exit future is fired,
-		// but we need to keep holding a reference to the global telemetry guard
-		// and drop the runtime first.
-		let _telemetry = service.telemetry();
-
-		// we hold a reference to the base path so if the base path is a temporary directory it will
-		// not be deleted before the tokio runtime finish to clean up
-		let _base_path = service.base_path();
+	fn run_service_until_exit(
+		mut self,
+		initialise: impl FnOnce(Configuration) -> sc_service::error::Result<KeepAliveServiceComponents>,
+	) -> Result<()> {
+		let KeepAliveServiceComponents {
+			task_manager,
+			// We need to keep a hold of these until the tasks are finished.
+			other: _other,
```

Comment:
What I was trying to solve in #6306 is that we have to do this thing of holding a reference to something we don't want to be dropped right away. This is one case here but when I write test I have the same special handling to do.

One way to solve this with your new implementation here would be to make TaskManager *not* implement Future (so it won't be consumed) but have a "`fn future(&self)`" that would return a future as you can see [here](https://github.com/paritytech/substrate/pull/6306/files#diff-8f956a9810426119c8963e0eb8bd2b74R269) with Service having a `fn future(&self)`.

By doing that, `task_manager` will not dropped. We can then call `terminate()` on it when the future is resolved and things need to be stopped. Like [here](https://github.com/paritytech/substrate/pull/6306/files#diff-8f956a9810426119c8963e0eb8bd2b74R274).

Now the problem is that we still need to keep those objects somewhere. Obviously there is no Service anymore to hold it for us. So I guess the TaskManager itself could do it for us (in other words: move the `other` field to `TaskManager`). Again, it doesn't need to know what's inside this `other` object, it just needs to hold it (Any).

So: instead of having a simple TaskManager. TaskManager would know that during termination, some things need to be hold and only when the TaskManager is dropped it would also drop the objects it was holding.

(Bonus: I think this will be a nice improvement also because the struct KeepAliveServiceComponents feel kinda unnecessary. It could be replaced by a tuple it wouldn't change anything.)

If you do this: it will fixes #6306 and `run_service_until_exit` will look much more stupid (which is great):

```rust
fn run_service_until_exit(self, initialized: ...) -> Result<()> {
    let task_manager = initialize(self.config)?;
    self.tokio_runtime.block_on(main(task_manager.future().fuse()));
    task_manager.terminate();
    drop(task_manager);
}
```

Reply:
(I think you will need to wrap the holding stuff in `Arc` to keep references of them)